mm2src/crypto/src/hw_client.rs

```
@@ -0,0 +1,94 @@
+            TrezorError::TransportNotSupported { transport } => HwError::TransportNotSupported { transport },
+            TrezorError::ErrorRequestingAccessPermission(_) => HwError::NoTrezorDeviceAvailable,
+            TrezorError::DeviceDisconnected => HwError::DeviceDisconnected,
+            TrezorError::UnderlyingError(_) => HwError::UnderlyingError(error),
+            TrezorError::ProtocolError(_) | TrezorError::UnexpectedMessageType(_) => HwError::Internal(error),
+            // TODO handle the failure correctly later
+            TrezorError::Failure(_) => HwError::Failure(error),
+            TrezorError::UnexpectedInteractionRequest(req) => HwError::UnexpectedUserInteractionRequest(req),
+            TrezorError::Internal(_) => HwError::Internal(error),
+        }
+    }
+}
+
+impl From<Bip32Error> for HwError {
+    fn from(e: Bip32Error) -> Self { HwError::InvalidXpub(e) }
+}
+
+#[derive(Clone)]
+pub enum HwClient {
+    Trezor(TrezorClient),
+}
+
+impl From<TrezorClient> for HwClient {
+    fn from(trezor: TrezorClient) -> Self { HwClient::Trezor(trezor) }
+}
+
+#[derive(Clone, Copy, Deserialize)]
+pub enum HwWalletType {
+    Trezor,
+}
+
+impl HwClient {
+    #[cfg(target_arch = "wasm32")]
+    pub async fn trezor() -> HwResult<TrezorClient> {
+        let mut devices = trezor::transport::webusb::find_devices().await?;
+        if devices.available.is_empty() {
+            return MmError::err(HwError::NoTrezorDeviceAvailable);
+        }
+        let device = devices.available.remove(0);
+        let transport = device.connect().await?;
+        let trezor = TrezorClient::init(transport).await?;
+        Ok(trezor)
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    pub async fn trezor() -> HwResult<TrezorClient> {
+        let mut devices = trezor::transport::usb::find_devices()?;
+        if devices.is_empty() {
+            return MmError::err(HwError::NoTrezorDeviceAvailable);
+        }
+        let device = devices.remove(0);
```

Maybe we can return an error that we don't support connecting more than one Trezor device at the same time if `devices.len() > 1`, I know this is a very rare case but we should cover it just in case.

---

Hmm, interesting idea, thanks! As for now, we can just return an error, but later we should ask the user to choose which device he wants to use.

---

Done, could you please take a look?